packages/wasm/crate/src/planner.rs

```diff
@@ -1,591 +1,389 @@
-    /// the view service when the plan is [`finish`](Planner::finish)ed.
-    pub fn spend(&mut self, note: Note, position: tct::Position) -> &mut Self {
-        let spend = SpendPlan::new(&mut self.rng, note, position).into();
-        self.action(spend);
-        self
+    fn balance_with_fee_estimate(&self, gas_prices: &GasPrices, fee_tier: &FeeTier) -> Balance {
+        self.balance() - self.fee_estimate(gas_prices, fee_tier).0
     }
 
-    /// Open a liquidity position in the order book.
-    pub fn position_open(&mut self, position: Position) -> &mut Self {
-        self.action(ActionPlan::PositionOpen(PositionOpen { position }));
-        self
+    fn refresh_change(&mut self, change_address: Address) {
+        self.change_outputs = BTreeMap::new();
+        // For each "provided" balance component, create a change note.
+        for value in self.balance().provided() {
+            self.change_outputs.insert(
+                value.asset_id,
+                OutputPlan::new(&mut OsRng, value, change_address),
+            );
+        }
     }
 
-    /// Close a liquidity position in the order book.
-    pub fn position_close(&mut self, position_id: position::Id) -> &mut Self {
-        self.action(ActionPlan::PositionClose(PositionClose { position_id }));
-        self
+    fn adjust_change_for_fee(&mut self, fee: Fee) {
+        self.change_outputs.entry(fee.0.asset_id).and_modify(|e| {
+            e.value.amount = e.value.amount.saturating_sub(&fee.0.amount);
+        });
     }
+}
 
-    /// Withdraw a liquidity position in the order book.
-    ///
-    /// Note: Currently this only supports an initial withdrawal from Closed, with no rewards.
-    pub fn position_withdraw(
-        &mut self,
-        position_id: position::Id,
-        reserves: Reserves,
-        pair: TradingPair,
-    ) -> &mut Self {
-        self.action(ActionPlan::PositionWithdraw(PositionWithdrawPlan {
-            reserves,
-            position_id,
-            pair,
-            sequence: 0,
-            rewards: Vec::new(),
-        }));
-        self
+/// Process a `TransactionPlannerRequest`, returning a `TransactionPlan`
+#[wasm_bindgen]
+pub async fn plan_transaction(
+    idb_constants: JsValue,
+    request: JsValue,
+    // TODO: use fvk to infer from source?
+    change_address: JsValue,
+    // all of these params should be removed
+    // TODO: read directly from IDB
+    fmd_params: JsValue,
+    // TODO: read directly from IDB
+    sct_params: JsValue,
+    // TODO: read directly from IDB
+    gas_prices: JsValue,
+    // TODO: read directly from IDB
+    chain_id: JsValue,
+) -> WasmResult<JsValue> {
+    utils::set_panic_hook();
+
+    let request: TransactionPlannerRequest = serde_wasm_bindgen::from_value(request)?;
+    // TODO: should we take an FVK to derive this from the data in the request?
+    let change_address: Address = serde_wasm_bindgen::from_value(change_address)?;
```

Comment:
I think it'd be a better API if we passed an FVK instead of the change address, but I wasn't sure how to do that and the changes were quite large.